### ISSUE TITLE
How to add batch normalization layer in the custom policy network

### ISSUE THREAD

**Issue by dongfangliu** [hill-a/stable-baselines#874]
I found it could be very difficult to not touch the source code when I am trying to add batch normalization layer. Because it requires an extra `training` tag and add initialization to the compute graph.

If we could add layers straightly as we did in keras, it will reduce pain. 

Is it any recommended implementation for batch normalization?

**Comment by Miffyli** [comment]
Related issue #424

[This reply] on that issue covers parts of the answer (and some recommendations why batch norm why not be the best idea for RL). As for the implementation: You have to modify the algorithm to have this `training` label and pass it along to correct initializations. There is no premade solution for this, and the linked reply covers why Keras layers functions might not work here.

**Comment by dongfangliu** [comment]
@Miffyli  Thank you.
I look through the referenced issue, but there are no info about why batch norm is not recommended. So could you just reply here directly, i am looking forward to your reply.

I checked the difference between batch norm and layer norm. I see that layer norm might be better for RNN. But i do not know why batch norm is not recommended.

**Comment by Miffyli** [comment]
The distribution of inputs changes as RL training proceeds forwards (you get to new states), so the normalization used for previous steps is not valid any more. Bigger nets do not provide large improvements to performance (see e.g. [IMPALA paper](https://arxiv.org/abs/1802.01561)), and even the larger ones used (3-layer resnet) do not need batch norm to be trained. I would not use larger nets before I have results with smaller net (e.g. the DQN's convolutional network).

**Comment by dongfangliu** [comment]
@Miffyli  Sounds reasonable. Thanks for your detailed explanation and references.

**Comment by Miffyli** [comment]
PS: Sorry for not providing any hard evidence ^^. This is something somebody should try out with bunch of environments and report in a workshop or so.

**Comment by dongfangliu** [comment]
@Miffyli But why layer norm could be useful, still distribution of states is changing

**Comment by Miffyli** [comment]
@dongfangliu 

Ah, my bad! I referenced a completely wrong issue. I was supposed to refer to this one: #373 

A [linked paper](https://arxiv.org/pdf/1812.02341.pdf) there shows how batch normalization can help with generalization to unseen levels.

**Comment by dongfangliu** [comment]
Thanks!

**Comment by zkx741481546** [comment]
> The distribution of inputs changes as RL training proceeds forwards (you get to new states), so the normalization used for previous steps is not valid any more. Bigger nets do not provide large improvements to performance (see e.g. [IMPALA paper](https://arxiv.org/abs/1802.01561)), and even the larger ones used (3-layer resnet) do not need batch norm to be trained. I would not use larger nets before I have results with smaller net (e.g. the DQN's convolutional network).

That isn't always right in certain RL problem, for example, stock trading. Some environments are driven by given data, whose input distribution won't change as RL training. 

Also, OpenAI has already proved BN can help RL model to against overfitting, training very deep network is not the only use of BN. (paper link: `https://arxiv.org/pdf/1812.02341.pdf`) 

So, I believe it's necessary to support BN operation in the future.

**Comment by Miffyli** [comment]
This could be added later on to [stable-baselines3](https://github.com/DLR-RM/stable-baselines3) (and should also be easier to add as a custom policy). This repository nowadays will mainly be updated with bug fixes and small tweaks.
